### PR TITLE
fix: randomx memory usage

### LINKED ISSUE #3104: RandomX VM cannot be created with less than 2GB memory
**Describe the bug**
Running the base node with less than 2080MB memory results in this error, a failure to create the randomx dataset. Even with randomx default flags, which should be able to create the VM in only 256MB memory. 

#3103

Investigation into [tari-project/randomx-rs ](/tari-project/randomx-rs) and [tevador/randomx](/tevador/randomx) are required to establish why the default flags still attempt to create the dataset and VM with full memory requirements of randomx 

```
2021-07-18 08:42:12.426544710 [c::pow::randomx_factory] WARN  Error initializing randomx dataset with flags FLAG_HARD_AES | FLAG_JIT | FLAG_ARGON2_SSSE3 | FLAG_ARGON2_AVX2 | FLAG_ARGON2. Problem creating the RandomX object:Could not allocate dataset. Fallback to default flags
2021-07-18 08:42:12.436775657 [c::bn::header_sync] WARN  Block header validation failed: Error in merge mine data:RandomX error: Problem creating the RandomX object:Could not allocate dataset
```

**To Reproduce**
1. Run weatherwax base node in a VM with less than 2GB memory 
2. Wait for headers to start syncing 
3. Observe error to create randomx dataset 

**Expected behavior**

RandomX VM should be started in light mode and be able to verify headers with only 256MB memory

**Desktop (please complete the following information):**
 - OS & Version: [e.g. iOS 10.2.1]
 ubuntu 20.04

**Additional context**

### PR DESCRIPTION
**Pull request by StriderDM**
Description
Removed dataset from RandomXVMInstance, light mode only requires a cache, this fixes a bug where light mode will consume more than 256MB of memory per RandomXVMInstance.

Made RandomXDataset optional for RandomXVMInstance.

resolves #3104 
closes #3103 

Motivation and Context
Overall memory overhead reduction

How Has This Been Tested?
Manually in a constrained vagrant instance

Screen Shot 2021-09-04 at 12 45 40 AM

### CODE REVIEW

**Comment by delta1** [comment]
Nice one! Will test

**Review by sdbondi — APPROVED** [review]
Nice - utAck. 

EDIT: How are validation times impacted?

File: base_layer/core/src/proof_of_work/randomx_factory.rs [hunk 1]
```diff
@@ -11,35 +11,43 @@ const LOG_TARGET: &str = "c::pow::randomx_factory";
 
 #[derive(Clone)]
 pub struct RandomXVMInstance {
-    // Note: If the cache and dataset drops, the vm will be wonky, so have to store all
-    // three for now
-    instance: Arc<Mutex<(RandomXVM, RandomXCache, RandomXDataset)>>,
+    // Note: If a cache and dataset (if assigned) allocated to the VM drops, the VM will crash.
+    // The cache and dataset for the VM need to be stored together with it since they are not
+    // mix and match.
+    instance: Arc<Mutex<(RandomXVM, RandomXCache, Option<RandomXDataset>)>>,
```
Comment (stringhandler) [inline base_layer/core/src/proof_of_work/randomx_factory.rs:17]: Why keep the dataset in here if it's always set to `None`?

**Review by stringhandler — APPROVED** [review]
Nice, one question though...

**Pull request merged**